packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -281,22 +281,26 @@ abstract class ImageProvider<T> {
       );
     }
 
-    /// `obtainKey` can throw both sync and async errors.
-    /// `catchError` handles cases where async errors are thrown and the outer try block is for sync errors.
-    ///
-    /// `onError` callback on [ImageCache] handles the cases where `obtainKey` is a sync future and `load` throws.
+    // `obtainKey` can throw both sync and async errors.
+    // `catchError` handles cases where async errors are thrown and the try block is for sync errors.
+    //
+    // `onError` callback on [ImageCache] handles the cases where `obtainKey` is a sync future and `load` throws.
+    Future<T> key;
     try {
-      obtainKey(configuration).then<void>((T key) {
-        obtainedKey = key;
-        final ImageStreamCompleter completer = PaintingBinding.instance
-            .imageCache.putIfAbsent(key, () => load(key), onError: handleError);
-        if (completer != null) {
-          stream.setCompleter(completer);
-        }
-      }).catchError(handleError);
+      key = obtainKey(configuration);
     } catch (error, stackTrace) {
       handleError(error, stackTrace);
```

Comment:
If you catch an error here `key` will be null, right?

Reply:
🤕 yeah, sorry will fix this

Reply:
Please add a test case which exercises this new code